docs/lesson-plans/week-4.md

```diff
@@ -0,0 +1,623 @@
+[JSON website](https://www.json.org/json-en.html).
+
+The JSON data format consists of three main data types; the `value`, `array`, and `object` types.
+
+##### Values
+
+Values are very simple, and there are only about seven main types: strings, numbers, objects, arrays, `true`, `false`,
+and `null`. 
+
+##### Arrays
+
+Arrays are a collection of `value`s commonly also referred to a vector, list, or sequence. See examples below:
+
+```json
+[]
+
+["a", "b", "c", "d", "e"]
+
+[[], ["a"], ["b", "c"]]
+
+[1, 2, ["a"], "b", true, false, null, [[[[]]]]]
+```
+
+##### Objects
+
+Objects are similar to arrays, however, they have keys which are assigned to each value. They're also commonly known as  
+serialized versions of an object, record, struct, dictionary, hash (the data type kind), map, hash table, keyed list, or
+associative array if you've heard of any of those before.
+
+>Serialization refers to the act of taking data from a data type or object state and turning it into a format that can
+>be stored/transmitted/more easily read and reconstructed later.
+
+>Deserialization refers to taking serialized data and turning it back into its original state.
+
+Here are some examples:
+
+```json
+{}
+
+{
+  "simple": "object"
+}
+
+{
+  "key": "value",
+  "other key": 5,
+  "nested object": {
+    "key": true,
+    "value": [1, 2, "a", { "hello": "world" }]
+  }
+}
+```
+
+#### Request Types Overview
+
+In HTTP (and as mentioned previously) there are multiple different request types. A lot of them will be covered in the 
+following lesson, so we'll just have a high level overview here.
+
+Each type of request is supposed to represent a different operation in the idea of [Create, Read, Update, Delete 
+(CRUD)](https://en.wikipedia.org/wiki/Create,_read,_update_and_delete).
+There are a couple outside of those, but in most cases, you'll just stick with those operations.
+
+##### GET
+
+The `GET` method is one of the most commonly used request methods (along with `POST`). It corresponds to the read 
+operation of CRUD, so it's usually used when you're trying to just get information from a server. In fact, whenever you
+load a website using a web browser, you're sending `GET` requests to the website you want to access. 
+
+### What is REST
+
+Representational State Transfer (REST) is a set of constraints for building web services. RESTful web services, or 
+web applications that conform to REST techniques require the following constraints.
+
+- *Server-Client Architecture*: The server-client architecture discussed earlier.
+- *Stateless*: the server can't store any client context on the server, so clients are the only ones that know about 
+their session state and contain all information necessary to make a request. If the server needs to know any of the 
+client context (like authentication information), it should be stored in a database.
+- *Cacheable*: Responses can be marked as cacheable or non-cacheable so that optimizations can be performed for 
```

Comment:
What does cacheable / non-cacheable mean?

Reply:
^

Reply:
Could use savable and non-savable instead

Reply:
switched it do Savable (Cacheable)